### PR TITLE
Poprouting plugin for OLSRd

### PR DESCRIPTION
**Pull request by gabri94**
See #38.
I squashed everything and re-done the pull request on the latest commits.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -94,7 +94,7 @@ endif
 cfgparser:	$(CFGDEPS) src/builddata.o
 		$(MAKECMDPREFIX)$(MAKECMD) -C $(CFGDIR)
 
-switch:		
+switch:
```
Comment (fhuberts) [inline Makefile]: remove this  change

File: Makefile.inc [hunk 2]
```diff
@@ -48,7 +48,7 @@
 #
 
 # activate debugging with 1 or deactivate with 0
-DEBUG ?= 1
+DEBUG ?= 0
```
Comment (fhuberts) [inline Makefile.inc]: remove this change

File: Makefile.inc [hunk 3]
```diff
@@ -57,7 +57,7 @@ NOSTRIP ?= 0
 NO_DEBUG_MESSAGES ?= 0
 
 # the optimize option to be set for gcc
-OPTIMIZE ?= 
+OPTIMIZE ?= -O3
```
Comment (fhuberts) [inline Makefile.inc]: remove this change

File: lib/info/info_types.h [hunk 4]
```diff
@@ -113,8 +113,17 @@ typedef struct {
 #define SIW_NETJSON_NETWORK_COLLECTION   (1ULL << 19)
 #define SIW_NETJSON                      (SIW_NETJSON_NETWORK_ROUTES | SIW_NETJSON_NETWORK_GRAPH | SIW_NETJSON_DEVICE_CONFIGURATION | SIW_NETJSON_DEVICE_MONITORING | SIW_NETJSON_NETWORK_COLLECTION)
 
+/*poprouting */
```
Comment (fhuberts) [inline lib/info/info_types.h]: add a space before poprouting

File: lib/info/olsrd_info.c [hunk 5]
```diff
@@ -479,7 +484,7 @@ static void send_info(const char * req, bool add_headers, unsigned int send_what
   unsigned int send_index = 0;
   bool first_reply = false;
 
-  const char *content_type = functions->determine_mime_type ? functions->determine_mime_type(send_what) : "text/plain; charset=utf-8";
+  const char *content_type = functions->determine_mime_type ? functions->determine_mime_type(send_what) : "text/plain; charset=utf-8"; //"
```
Comment (fhuberts) [inline lib/info/olsrd_info.c]: remove this change

File: lib/info/olsrd_info.c [hunk 6]
```diff
@@ -526,6 +531,15 @@ static void send_info(const char * req, bool add_headers, unsigned int send_what
...
 
       send_info_from_table(&abuf, send_what, funcs, ARRAY_SIZE(funcs), &outputLength);
+    } else if(send_what & SIW_POPROUTING){
+      SiwLookupTableEntry funcs[] = {
+        { SIW_POPROUTING_TC               , functions->TcTimer           }, //
+        { SIW_POPROUTING_HELLO            , functions->HelloTimer        }, //
+        { SIW_POPROUTING_TC_MULT          , functions->TcTimerMult       }, //
+        { SIW_POPROUTING_HELLO_MULT       , functions->HelloTimerMult    } //
+      };
+      send_info_from_table(&abuf, send_what, funcs, ARRAY_SIZE(funcs), &outputLength);
```
Comment (fhuberts) [inline lib/info/olsrd_info.c]: add space before this line and remove it after this line

File: lib/info/info_types.h [hunk 7]
```diff
@@ -163,6 +172,11 @@ typedef struct {
     printer_generic deviceConfiguration;
     printer_generic deviceMonitoring;
     printer_generic networkCollection;
+    printer_generic TcTimer;
+    printer_generic HelloTimer;
+    printer_generic TcTimerMult;
+    printer_generic HelloTimerMult;
+
```
Comment (fhuberts) [inline lib/info/info_types.h]: remove this empty line

File: lib/info/info_types.h [hunk 8]
```diff
@@ -163,6 +172,11 @@ typedef struct {
     printer_generic deviceConfiguration;
     printer_generic deviceMonitoring;
     printer_generic networkCollection;
+    printer_generic TcTimer;
+    printer_generic HelloTimer;
+    printer_generic TcTimerMult;
+    printer_generic HelloTimerMult;
```
Comment (fhuberts) [inline lib/info/info_types.h]: function names should stick to proper casing, so TcTimer should be tcTimer, etc

File: lib/poprouting/Makefile [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+# Visit http://www.olsr.org for more information.
+#
+# If you find this software useful feel free to make a donation
+# to the project. For more information see the website or contact
+# the copyright holders.
+#
+
+OLSRD_PLUGIN =	true
+PLUGIN_NAME =	olsrd_poprouting
+PLUGIN_VER =	0.1
```
Comment (fhuberts) [inline lib/poprouting/Makefile]: please start at 1.0, any change after that should stick to semantic versioning

File: lib/poprouting/README_POPROUTING [hunk 10]
```diff
@@ -0,0 +1,38 @@
+============
+INTRODUCTION
+============
+
+The netjson plugin is not an info plugin, but it uses the info structure to allow communication.
```
Comment (fhuberts) [inline lib/poprouting/README_POPROUTING]: not netjson but poprouting

File: lib/poprouting/README_POPROUTING [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+* /TCTimerMult
+====================
+PLUGIN CONFIGURATION
+====================
+
+The plugin is configured with the generic info plugin configuration parameters.
+
+The port in the generic info plugin configuration is set to 2007.
+
+LoadPlugin "olsrd_poprouting.so.0.1"
```
Comment (fhuberts) [inline lib/poprouting/README_POPROUTING]: 1.0

File: lib/poprouting/README_POPROUTING [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+* /TCTimer
+* /HelloTimerMult
+* /TCTimerMult
+====================
+PLUGIN CONFIGURATION
+====================
+
+The plugin is configured with the generic info plugin configuration parameters.
+
+The port in the generic info plugin configuration is set to 2007.
```
Comment (fhuberts) [inline lib/poprouting/README_POPROUTING]: pick a different port. I think 2007 clashes with an existing plugin

**Comment by gabri94** [comment]
I did the requested changes.

**Comment by gabri94** [comment]
Any news about the review?

**Comment by gabri94** [comment]
Yes, I pushed them with -f to avoid multiple commits.

**Comment by fhuberts** [comment]
your push doesn't show up in github, please check

File: lib/poprouting/Makefile [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+# Visit http://www.olsr.org for more information.
+#
+# If you find this software useful feel free to make a donation
+# to the project. For more information see the website or contact
+# the copyright holders.
+#
+
+OLSRD_PLUGIN =	true
+PLUGIN_NAME =	olsrd_poprouting
+PLUGIN_VER =	1.0
```
Comment (gabri94) [inline lib/poprouting/Makefile]: I see the changes

File: lib/info/info_types.h [hunk 14]
```diff
@@ -163,6 +172,11 @@ typedef struct {
     printer_generic deviceConfiguration;
     printer_generic deviceMonitoring;
     printer_generic networkCollection;
+    printer_generic tcTimer;
```
Comment (gabri94) [inline lib/info/info_types.h]: same here

File: Makefile.inc [hunk 15]
```diff
@@ -57,7 +57,7 @@ NOSTRIP ?= 0
 NO_DEBUG_MESSAGES ?= 0
 
 # the optimize option to be set for gcc
-OPTIMIZE ?= 
+OPTIMIZE ?=
```
Comment (gabri94) [inline Makefile.inc]: Here i removed the -O3

File: lib/info/info_types.h [hunk 16]
```diff
@@ -113,8 +113,17 @@ typedef struct {
 #define SIW_NETJSON_NETWORK_COLLECTION   (1ULL << 19)
 #define SIW_NETJSON                      (SIW_NETJSON_NETWORK_ROUTES | SIW_NETJSON_NETWORK_GRAPH | SIW_NETJSON_DEVICE_CONFIGURATION | SIW_NETJSON_DEVICE_MONITORING | SIW_NETJSON_NETWORK_COLLECTION)
 
+/* poprouting */
```
Comment (gabri94) [inline lib/info/info_types.h]: Here i addedd the space

**Comment by fhuberts** [comment]
ok, I was checking the first change I asked you to do, which you didn't.
I'll review later

**Comment by gabri94** [comment]
I reverted the change in Makefile, now everything should be ok

File: lib/poprouting/src/olsrd_poprouting.c [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+    char response[1024];
+    sprintf(response, "hello:%.2f\n", (double)in->cnf->hello_params.emission_interval);
+    abuf_puts(abuf, response);
+    return;
+  }
+  while (in != NULL) {
+    olsr_printf(1, "(POPROUTING) Setting Hello Timer=%f for interface %s\n", (double)timer, in->name);
+    //olsr_change_timer(in->interf->hello_gen_timer, timer * MSEC_PER_SEC, 0, 1);
+    in->interf->hello_gen_timer->timer_period = timer * MSEC_PER_SEC;
+    in->interf->hello_gen_timer->timer_jitter_pct = POPROUTING_JITTER; // Jitter to 5%
```
Comment (fhuberts) [inline lib/poprouting/src/olsrd_poprouting.c]: why do this? you overwrite the configured value. I don't like this

File: lib/poprouting/src/olsrd_poprouting.c [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+    in->cnf->hello_params.emission_interval = timer;
+    in->cnf->hello_params.validity_time = timer * hello_mult;
+    in->interf->valtimes.hello = reltime_to_me(in->cnf->hello_params.validity_time * MSEC_PER_SEC);
+    in->interf->hello_etime = in->cnf->hello_params.emission_interval * MSEC_PER_SEC;
+    in = in->next;
+  }
+  abuf_puts(abuf, "0\r\n");
```
Comment (fhuberts) [inline lib/poprouting/src/olsrd_poprouting.c]: no windows EOLs, use \n only

File: lib/poprouting/src/olsrd_poprouting.c [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+    abuf_appendf(abuf, "error: %s\n", httpStatusToReply(status));
+  }
+}
+
+void set_hello_timer(struct autobuf *abuf) {
+  struct olsr_if *in;
+  assert(abuf);
+  in = olsr_cnf->interfaces;
+  if(!timer){
+    char response[1024];
```
[the remark this answers is not in the document]
Comment (fhuberts) [reply lib/poprouting/src/olsrd_poprouting.c]: multiple places

File: lib/poprouting/src/olsrd_poprouting.c [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+    sprintf(response, "hello:%.2f\n", (double)in->cnf->hello_params.emission_interval);
+    abuf_puts(abuf, response);
+    return;
+  }
+  while (in != NULL) {
```
[the remark this answers is not in the document]
Comment (fhuberts) [reply lib/poprouting/src/olsrd_poprouting.c]: multiple places

File: lib/poprouting/src/olsrd_poprouting.c (hunk 18 above)
Comment (fhuberts) [reply lib/poprouting/src/olsrd_poprouting.c]: multiple places

File: lib/poprouting/src/olsrd_poprouting.c (hunk 17 above)
Comment (fhuberts) [reply lib/poprouting/src/olsrd_poprouting.c]: multiple places

File: lib/poprouting/src/olsrd_poprouting.c (hunk 19 above)
[the remark this answers is not in the document]
Comment (gabri94) [reply lib/poprouting/src/olsrd_poprouting.c]: I setted it to 20, it shoud be enough.

File: lib/poprouting/src/olsrd_poprouting.c [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+
+    default:
+      return false;
+  }
+  cmd = strtok(string, s);
+  s_timer = strtok(NULL, s);
+  if (s_timer)
+    timer = atof(s_timer);
+  else
+    timer = 0;
```
[the remark this answers is not in the document]
Comment (gabri94) [reply lib/poprouting/src/olsrd_poprouting.c]: fixed it.

File: lib/poprouting/README_POPROUTING [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+It can be also used to get the current values from the daemon.
+
+Please first read what's written in the file lib/info/README_INFO.
+
+
+==================
+SUPPORTED COMMANDS
+==================
+The commands can be specified with or without a float number after the =.
+In the former case the timer or the multiplier will be setted at that specific value.
```
[the remark this answers is not in the document]
Comment (gabri94) [reply lib/poprouting/README_POPROUTING]: corrected

File: lib/poprouting/src/olsrd_poprouting.c [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+#include "lq_plugin.h"
+#include "gateway.h"
+#include "gateway_costs.h"
+#include "olsrd_plugin.h"
+#include "info/info_types.h"
+#include "info/http_headers.h"
+#include "gateway_default_handler.h"
+#include "scheduler.h"
+
+float timer=0;
```
[the remark this answers is not in the document]
Comment (gabri94) [reply lib/poprouting/src/olsrd_poprouting.c]: I'd really avoid the use of globals but I can't see any other way without changing the info plugin code. Any suggestion?

File: lib/poprouting/src/olsrd_poprouting.c (hunk 17 above)
Comment (gabri94) [reply lib/poprouting/src/olsrd_poprouting.c]: i chose to change also the configuration in order to keep the configuration coherent with the timers.

File: lib/poprouting/src/olsrd_poprouting.c (hunk 18 above)
Comment (gabri94) [reply lib/poprouting/src/olsrd_poprouting.c]: fixed it

File: lib/poprouting/src/olsrd_poprouting.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (fhuberts) [reply lib/poprouting/src/olsrd_poprouting.c]: well...
1: disable composite commands, or
2: have 1 global per command, that way the last one overrules a previous value of the same command

File: lib/poprouting/src/olsrd_plugin.c [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+int olsrd_plugin_interface_version(void) {
+  return PLUGIN_INTERFACE_VERSION;
+}
+
+static const struct olsrd_plugin_parameters plugin_parameters[] = { //
+    //
+        INFO_PLUGIN_CONFIG_PLUGIN_PARAMETERS(config), //
+        { .name = "vtime", .set_plugin_parameter = &set_plugin_boolean, .data = &vtime } //
```
[the remark this answers is not in the document]
Comment (gabri94) [reply lib/poprouting/src/olsrd_plugin.c]: i removed all the references and also the function:
olsrd_get_plugin_parameters()

File: lib/info/info_types.h [hunk 25]
```diff
@@ -113,8 +113,14 @@ typedef struct {
...
+#define SIW_POPROUTING_HELLO             (1ULL << 20)
+#define SIW_POPROUTING_TC                (1ULL << 21)
+#define SIW_POPROUTING_HELLO_MULT        (1ULL << 22)
+#define SIW_POPROUTING_TC_MULT           (1ULL << 23)
+#define SIW_POPROUTING                   (SIW_POPROUTING_HELLO | SIW_POPROUTING_TC | SIW_POPROUTING_HELLO_MULT | SIW_POPROUTING_TC_MULT)
 /* everything */
```
[the remark this answers is not in the document]
Comment (gabri94) [reply lib/info/info_types.h]: done

File: lib/info/info_types.h [hunk 26]
```diff
@@ -163,6 +169,11 @@ typedef struct {
     printer_generic deviceConfiguration;
     printer_generic deviceMonitoring;
     printer_generic networkCollection;
+    printer_generic tcTimer;
```
[the remark this answers is not in the document]
Comment (gabri94) [reply lib/info/info_types.h]: done

File: lib/info/info_types.h [hunk 27]
```diff
@@ -163,6 +169,11 @@ typedef struct {
...
+    printer_generic helloTimer;
+    printer_generic tcTimerMult;
+    printer_generic helloTimerMult;
+
```
[the remark this answers is not in the document]
Comment (gabri94) [reply lib/info/info_types.h]: done

File: lib/info/olsrd_info.c [hunk 28]
```diff
@@ -526,6 +531,15 @@ static void send_info(const char * req, bool add_headers, unsigned int send_what
...
 
       send_info_from_table(&abuf, send_what, funcs, ARRAY_SIZE(funcs), &outputLength);
+    } else if(send_what & SIW_POPROUTING){
+      SiwLookupTableEntry funcs[] = {
+        { SIW_POPROUTING_TC               , functions->tcTimer           }, //
+        { SIW_POPROUTING_HELLO            , functions->helloTimer        }, //
+        { SIW_POPROUTING_TC_MULT          , functions->tcTimerMult       }, //
+        { SIW_POPROUTING_HELLO_MULT       , functions->helloTimerMult    } //
+      };
+      send_info_from_table(&abuf, send_what, funcs, ARRAY_SIZE(funcs), &outputLength);
```
[the remark this answers is not in the document]
Comment (gabri94) [reply lib/info/olsrd_info.c]: corrected

File: lib/poprouting/src/olsrd_poprouting.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (gabri94) [reply lib/poprouting/src/olsrd_poprouting.c]: Ok, In the end globals cannot be avoided at all. 
I disabled the composite commands

**Comment by gabri94** [comment]
I'd have fixed everything.

File: lib/poprouting/src/olsrd_plugin.c (hunk 24 above)
[the remark this answers is not in the document]
Comment (fhuberts) [reply lib/poprouting/src/olsrd_plugin.c]: IIRC you should not remove that function. did you compile and run-test?

**Comment by fhuberts** [comment]
also, please don't amend the commit, make a new commit on top. amending makes it much more difficult for me to review since I have to do it all again.

put the plugin_parameters back and then we're about ready to merge

**Comment by gabri94** [comment]
Yes I know that is worse for you, but since at the beginning you asked me to squash all my commit into a single one, I kept doing it.

File: lib/poprouting/src/olsrd_plugin.c [hunk 29]
```diff
@@ -120,6 +120,16 @@ int olsrd_plugin_interface_version(void) {
   return PLUGIN_INTERFACE_VERSION;
 }
 
+static const struct olsrd_plugin_parameters plugin_parameters[] = { //
+    //
+        INFO_PLUGIN_CONFIG_PLUGIN_PARAMETERS(config)//
```
Comment (gabri94) [inline lib/poprouting/src/olsrd_plugin.c]: should I keep this line?

Comment (fhuberts) [reply lib/poprouting/src/olsrd_plugin.c]: yes

**Comment by gabri94** [comment]
It should be ready then

**Pull request merged**